### PR TITLE
Duplicate block field

### LINKED ISSUE #315: "Copy" or "Duplicate" block field action
It would be helpful to be able to quickly copy/duplicate a block field..

### PR DESCRIPTION
**Pull request by lukecarbis**
Wow! This was a lot more complicated than I thought it would be.

This isn't yet working for repeater fields (but it is for sub-fields).

This PR also includes a whole heap of JS coding standard fixes.

Closes #315.

### CODE REVIEW

**Comment by RobStino** [comment]
This is awesome!
Tested and the UX is great!

**Review by RobStino — APPROVED** [review]

File: js/admin.block-post.js [hunk 1]
```diff
@@ -4,24 +4,23 @@
  * @package   Block_Lab
  * @copyright Copyright(c) 2019, Block Lab
  * @license http://opensource.org/licenses/GPL-2.0 GNU General Public License, version 2 (GPL-2.0)
- *
- * Globals wp, blockLab
  */
 
-(function( $ ) {
+/* global blockLab, jQuery */
```
Comment (kienstra) [inline js/admin.block-post.js:9]: Wow, this is great. It looks like this now passes `eslint`!

What do you think of adding it to the `eslint` configuration?

```diff
diff --git a/package.json b/package.json
index c5cd510..6c7cf1a 100644
--- a/package.json
+++ b/package.json
@@ -11,7 +11,7 @@
     "dev": "cross-env BABEL_ENV=default webpack --watch",
     "build": "cross-env BABEL_ENV=default NODE_ENV=production webpack",
     "lint": "npm-run-all --parallel lint:*",
-    "lint:js": "eslint js/blocks",
+    "lint:js": "eslint js",
     "lint:js:fix": "eslint js/blocks --fix",
     "lint:php": "vendor/bin/phpcs",
     "lint:php:fix": "vendor/bin/phpcbf"
```

Also, creating an `.eslintignore` with the following will prevent it from running on the compiled files:

```js
js/editor.blocks.js
js/scripts.js
```

**Comment by kienstra** [comment]
**Looks Good**

Hi @lukecarbis,
This looks good, and duplication worked really well.

# Initial Field
initial-text-field

# Duplicate

dup-text

All of the fields that I tested had all of their values duplicated. Including Text, Post, and Checkbox.

**Review by kienstra — APPROVED** [review]

**Comment by kienstra** [comment]
**Side Note**

We might consider refactoring this Edit Block UI in React at some point (though not necessarily re-designing it much):

ed-block-ui

[js/admin.block-post.js](https://github.com/getblocklab/block-lab/pull/438/files#diff-d5c1a313545073ab02daba7f951160f3R4) had been long before, and is now 458 lines.

The DOM manipulation in that file gets really complex, and it's going to be hard to maintain. And that file's probably not going to get smaller 😄

**Comment by lukecarbis** [comment]
@kienstra @RobStino Requesting a re-review. Some big changes made in the last couple of commits.

In particular, please review duplication repeater fields!

**Comment by RobStino** [comment]
Hey all,
Tested and this works really well!
My testing included:
- Duplicating a field
- Duplicating a whole repeater field with subfields
- Duplicating a sub-field
- Duplicating a duplicated sub-field within a duplicated repeater field :P
Basically I tried all the variations of everything and noted that:
- Duplications always worked
- The generated field name was always unique

Great work @lukecarbis This is a really useful feature!

File: package.json [hunk 2]
```diff
@@ -11,8 +11,8 @@
     "dev": "cross-env BABEL_ENV=default webpack --watch",
     "build": "cross-env BABEL_ENV=default NODE_ENV=production webpack",
     "lint": "npm-run-all --parallel lint:*",
-    "lint:js": "eslint js/blocks",
-    "lint:js:fix": "eslint js/blocks --fix",
+    "lint:js": "eslint js",
+    "lint:js:fix": "eslint js --fix",
```
Comment (kienstra) [inline package.json]: Great idea

**Review by kienstra — APPROVED** [review]
**Great work!**

Hi @lukecarbis,
Wow, this works well, even duplicating entire repeaters.

Duplicating single fields and repeater rows also work well.

It might be good to refactor the 'Edit Block' screen with React soon. 

`admin.block-post.js` is over 500 lines, and will only get harder to refactor. Though it's worked really well, very few bugs.

**Comment by lukecarbis** [comment]
You guys. Thanks for testing, but you missed a critical bug!

Repeater child fields aren't saving. Will take a look at it asap.

**Comment by lukecarbis** [comment]
The bug is occurring because we're assigning the sub-fields their parent's UID instead of their parent's slug.

Actually, this made me realise that saving sub-fields with the parents name is a bad idea! It surfaced a new bug: If you create a repeater, add sub fields, then go back and change the repeater's name, then the sub fields will be deleted on save.

I'm going to refactor the sub fields to use the parent UID instead of the parent slug / name.

**Comment by lukecarbis** [comment]
@kienstra and @RobStino – please re-review? Especially testing:

- Do sub-rows from pre-existing blocks still save without making changes?
- Do sub-rows from pre-existing blocks still save _with_ making changes?
- Do new sub-rows still save?
- Can you break the auto-incrementing field names by adding rows, sub rows, and duplicating things?

**Comment by RobStino** [comment]
Just testing this now, and looks like I hit a limit for the incremental numbering for duplicating fields.
image
Is this intentional?

**Comment by lukecarbis** [comment]
@RobStino Great pickup. Fixed in https://github.com/getblocklab/block-lab/pull/438/commits/a650df07e7dc5163ac808de08a628767621ecaf1.

**Comment by RobStino** [comment]
Yep, that fixes it.

**Comment by kienstra** [comment]
|  Do sub-rows from pre-existing blocks still save without making changes? | yes  |
|---|---|
| Do sub-rows from pre-existing blocks still save with making changes?  |  yes  |
|  Do new sub-rows still save? |  yes  |


new-sub-row

I'm working on this:
>Can you break the auto-incrementing field names by adding rows, sub rows, and duplicating things?

**Comment by kienstra** [comment]
>Can you break the auto-incrementing field names by adding rows, sub rows, and duplicating things?

No, I haven't been able to break this 😄

**Comment by kienstra** [comment]
Pending any testing you'd like, I think the code is ready to merge.

**Pull request merged**